DirectProgramming/DPC++FPGA/ReferenceDesigns/gunzip/README.md

```diff
@@ -0,0 +1,281 @@
+
+#### LZ77 Decoder
+The input to the LZ77 decoder is a stream of commands that are either an 8-bit literal or a {length, distance} pair; the output of the Huffman decoder. The output is a stream of literals (8-bit characters). The LZ77 decoder keeps a 32K *history buffer*; a history of the last 32K literals it has streamed to the output. 
+
+If the incoming command is an 8-bit literal, the LZ77 decoder simply forwards the literal to the output stream and tracks it in the history buffer. For a {length, distance} pair, the decoder goes back `distance` literals in the history buffer and streams `length` of them to the output and back into the history buffer.
+
+The LZ77 decoder can be configured to read `N` elements from the history buffer per cycle, where `N` is a compile time constant (NOTE: in the source code `N` is `literals_per_cycle`). Therefore, the output of the LZ77 decoder is actually an array of literals where the number of valid literals is in the range `[1, N]`, which is indicated by the `valid_count` variable.
+
+The image below illustrates the multi-element LZ77 decoder. To read multiple elements from the history buffer at once, the history buffer is cyclically partitioned across `N` buffers. For example, if `N=4`, then the order in which literals are written to the history buffers is `0, 1, 2, 3, 0, 1, 2, 3, ...`. The *Current History Buffer Index* indicates which of the `N` history buffers is the *next* to be written to. When `N` (or less) elements come in the input, the elements are shuffled to the `N` buffers in the correct order. For example if *Current History Buffer Index* is 2, then the `N=4` incoming elements should be written to history buffers `{2, 3, 0, 1}`, respectively. In general, the order is `{Current History Buffer Index, (Current History Buffer Index + 1) % N, ..., (Current History Buffer Index + N - 1) % N}`.
+
+<img src="lz77_decoder.png" alt="lz77_decoder" width="700"/>
+
+When reading out of the history buffers, the *Current History Buffer Index* is again used to shuffle the output of the `N` buffers so that they are streamed out in the correct order. The following example illustrates reading the history buffer for `N=4`.
+
+Let's assume that to start the LZ77 decoder received 14 consecutive literals and wrote them to the output and history buffers in cyclical order (i.e. to buffers `0, 1, 2, 3, 0, 1, 2, 3, ...`).
+```
+Characters:    E T H O M E P H O N E
```

Comment:
Really?  ET HOME PHONE, not ET PHONE HOME?

Reply:
https://www.youtube.com/watch?v=iyFijjikkeM